spinalcordtoolbox/resampling.py

```diff
@@ -49,13 +49,21 @@ def resample_nib(image, new_size=None, new_size_type=None, image_dest=None, inte
     dict_interp = {'nn': 0, 'linear': 1, 'spline': 2}
 
     # If input is an Image object, create nibabel object from it
-    if type(image) == nib.nifti1.Nifti1Image:
+    if isinstance(image, nib.nifti1.Nifti1Image):
         img = image
-    elif type(image) == Image:
+    elif isinstance(image, Image):
         img = nib.nifti1.Nifti1Image(image.data, image.hdr.get_best_affine(), image.hdr)
     else:
         raise TypeError(f'Invalid image type: {type(image)}')
 
+    # convert to floating point if we're doing arithmetic interpolation
+    if interpolation != 'nn' and img.get_data_dtype().kind in 'biu':
+        original_dtype = img.get_data_dtype()
+        img = nib.nifti1.Nifti1Image(img.get_fdata(), img.affine, img.header)
+        img.set_data_dtype(img.dataobj.dtype)
+        logger.warning("Converting image from type '%s' to type '%s' for %s interpolation",
+                       original_dtype, img.get_data_dtype(), interpolation)
```

Comment:
Just out of curiosity, did you by chance look into whether there's a way to avoid this on the nibabel side of things? (i.e. passing an integer image into nibabel's resampling function, but somehow automatically returning a float data array?)

I ask because I'm curious where the actual type coercion is happening, and (if it's happening inside of nibabel), whether there's a more "official" way to avoid the coercion, so to speak.

Reply:
I remember spending time looking at this. Explorations brought me to scikit-image's rationale for converting output dtype when _it made sense_, see notably [this doc](https://scikit-image.org/docs/stable/user_guide/data_types.html#output-types). I'm not sure if nibabel is relying on scikit library though, but the rationale still holds, I think.

Reply:
I'll try and dig a bit right now using the data you provided in the issue. :)

Reply:
Quick investigation:

- `nibabel` is using [`scipy.ndimage.affine_transform`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.affine_transform.html) under the hood
- `affine_transform` has an `output` parameter with the following description:
    >   output : array or dtype, optional
    > 
    >   The array in which to place the output, or the dtype of the returned array. By default an array of the same dtype as input will be created.
- `nibabel` does not specify this parameter, nor expose it to the `resample_from_to` function.

I could petition `nibabel` to expose this parameter, but otherwise I think @mguaypaq's solution (changing the dtype of the input array) is the best we can do right now given the default behavior of `affine_transform`.

Reply:
Thanks for the investigation!